src/audio/copier/copier.c

```diff
@@ -122,46 +120,83 @@ static struct comp_dev *copier_new(const struct comp_driver *drv,
 		case ipc4_i2s_link_input_class:
 		case ipc4_alh_link_output_class:
 		case ipc4_alh_link_input_class:
-			if (copier_dai_create(dev, cd, &dev->ipc_config,
-					      copier, ipc_pipe->pipeline)) {
-				comp_cl_err(&comp_copier, "unable to create dai");
-				goto error_cd;
+			ret = copier_dai_create(dev, cd, &dev->ipc_config,
+						copier, ipc_pipe->pipeline);
+			if (ret < 0) {
+				comp_err(dev, "unable to create dai");
+				goto error;
 			}
 			break;
 #if CONFIG_IPC4_GATEWAY
 		case ipc4_ipc_output_class:
 		case ipc4_ipc_input_class:
-			if (copier_ipcgtw_create(dev, cd, &dev->ipc_config,
-						 copier, ipc_pipe->pipeline)) {
-				comp_cl_err(&comp_copier, "unable to create IPC gateway");
-				goto error_cd;
+			ret = copier_ipcgtw_create(dev, cd, &dev->ipc_config,
+						   copier, ipc_pipe->pipeline);
+			if (ret < 0) {
+				comp_err(dev, "unable to create IPC gateway");
+				goto error;
 			}
-
 			break;
 #endif
 		default:
-			comp_cl_err(&comp_copier, "unsupported dma type %x",
-				    (uint32_t)node_id.f.dma_type);
-			goto error_cd;
+			comp_err(dev, "unsupported dma type %x", (uint32_t)node_id.f.dma_type);
+			ret = -EINVAL;
+			goto error;
 		};
 
 		dev->direction_set = true;
 	}
 
 	dev->direction = cd->direction;
 	dev->state = COMP_STATE_READY;
-	return dev;
-
-error_cd:
-	rfree(cd);
+	return 0;
 error:
+	rfree(cd);
+	return ret;
+}
+
+static struct comp_dev *copier_new(const struct comp_driver *drv,
+				   const struct comp_ipc_config *config,
+				   const void *spec)
+{
+	struct module_config *dst;
+	struct comp_dev *dev;
+	struct processing_module *mod;
+	struct module_data *md;
+	int ret;
+
+	comp_cl_dbg(&comp_copier, "copier_new()");
+
+	mod = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*mod));
+	if (!mod)
+		return NULL;
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		goto error_dev;
+
+	dev->ipc_config = *config;
+
+	comp_set_drvdata(dev, mod);
+	md = &mod->priv;
+	mod->dev = dev;
+	dst = &md->cfg;
+	dst->init_data = spec;
```

Comment:
how about just `md->cfg.init_data = spec` and remove `dst`. Or even `mod->priv.cfg.init_data = spec` and remove both `md` and `dst`

Reply:
In assembly level, there is no difference at all for both implementation.

Reply:
@lyakh @lgirdwood , use md->cfg.init_data = spec to remove dst usage, mod->priv.cfg.init_data = spec this is a little long, I prefer use md->cfg.init_data = spec.

Reply:
> In assembly level, there is no difference at all for both implementation.

Sure, but we aren't programming in assembly. There are many ways to write code in C to generate the same assembly output. You can name your variables `a1`, `a2` etc. but that's not how we're programming

Reply:
fully agree, please review the changes, and let me know if you still have further comments.